### ISSUE TITLE
Model Update method does not respect validators

### ISSUE THREAD

**Issue by ruimgoncalves** [Automattic/mongoose#860]
I can use the model.Update method to update to any value regardless of schema constraints

For example if I have a Schema with a enum property, using update I can change to any value outside of the enum constraint, thus invalidating the set.

The expected behavior was to throw an error if the updated value did not validate.

I suggest an extra option to the update method to exit on the first error or not.

**Comment by vkarpov15** [comment]
@andrewholsted waiting on [mongodb driver 2.0](https://github.com/mongodb/node-mongodb-native) and mongodb server 2.8 to stabilize - that's hopefully happening this month, but can't really ship 4.0 without supporting latest version of mongodb and latest driver. See [my blog](http://thecodebarbarian.com/2014/12/19/mongoose-397) for a little more detail.

**Comment by zacharynevin** [comment]
Potentially dumb question by me, but why not take the data being updated and validate that against the schema before passing to mongodb?

Also, +1

**Comment by zhiqingchen** [comment]
mark & wait

**Comment by iamceege** [comment]
Can't wait until 3.9 is stable :)

**Comment by bunbutter** [comment]
Is this feature in yet? Eagerly waiting for 3.9.. Wen will it be available?

**Comment by vkarpov15** [comment]
You can `npm install mongoose@unstable` to get the unstable version. Current ETA for 4.0 is march 25 - a good place to check this is the [milestones page](https://github.com/LearnBoost/mongoose/milestones)

**Comment by Pathsofdesign** [comment]
Just ran into this issue last night. Glad I'm not the only one! I'm excited for this change to be released, thanks for all you do!

+1

**Comment by Weblors** [comment]
Don't know why this issue is marked closed. I'm still facing it.

**Comment by vkarpov15** [comment]
See [update validators section in validation docs](http://mongoosejs.com/docs/validation.html)

**Comment by m1cah** [comment]
I'm actually not sure that this works on enum validation for findOneAndUpdate(), even with runValidators set to true.

**Comment by vkarpov15** [comment]
@m1cah please provide example that demonstrates what you're trying to do. We do have tests for this and they pass...

**Comment by m1cah** [comment]
@vkarpov15 I believe this is a short example that demonstrates it: http://code.runnable.com/VYhGbVhereIYdbst/update-validation-enum-for-mongoose-and-databases

**Comment by vkarpov15** [comment]
Well one issue is that the example above is using a prehistoric version of mongoose:

```
root@runnable:~# head node_modules/mongoose/package.json                                                                                                                 
{                                                                                                                                                                         
  "name": "mongoose",                                                                                                                                                     
  "description": "Elegant MongoDB object modeling for Node.js",                                                                                                           
  "version": "3.6.14",                                                                                                                                                    
  "author": {                                                                                                                                                             
    "name": "Guillermo Rauch",                                                                                                                                            
    "email": "guillermo@learnboost.com"                                                                                                                                   
  },                                                                                                                                                                      
  "keywords": [                                                                                                                                                           
    "mongodb",  
```

Try upgrading to 4.x and it should work.

**Comment by nirio69510** [comment]
Are you sure it works with an enum and findOneAndUpdate method ? 
On mongoose 4.2.6 it seems to fail, I can set a bad value. 

Schema : 

    var UserSchema = new Schema({
        first_name: {
            type: String,
            required: true,
        },
        last_name: {
            type: String,
            required: true,
        },
        email: {
            type: String,
            unique: true,
            required: true,
            uniqueCaseInsensitive: true
        },
        embededData: [{
            type: {
                type: String,
                enum: ['value1', 'value2', 'value3']
                required: true
            }
        }]
    }, { strict: true });

FindOneAndUpdate method : 

    UserModel.findOneAndUpdate(
        {_id: uid}, 
        {$push: {embededData: data}}, 
        { runValidators: true }, function(err) {
    });

Then I can push ```embededData.type = 'Panda';```

**Comment by vkarpov15** [comment]
See [update validator docs](http://mongoosejs.com/docs/validation.html#update-validators) and #2933 - update validators don't run on `$push`, only on `$set` and `$unset`

**Comment by ajbraus** [comment]
+1

**Comment by rmacqueen** [comment]
As far as I can see, this still doesn't work for the case where you are using a custom validator on a field that is an array. 

For example, this code snippet: 

```
var mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/myTestDB');
 
var db = mongoose.connection;
 
db.on('error', function (err) {
console.log('connection error', err);
});
db.once('open', function () {
console.log('connected.');
});
 
var Schema = mongoose.Schema;
var userSchema = new Schema({
  _id : String,
  name : {
    type: [String],
    validate: {
        validator: function (str) {
            return str.length > 1
        }
    },
  }
});


var User = mongoose.model('User', userSchema);

User.findOneAndUpdate({"name": ["John", "Doe"]}, { 
  $setOnInsert: {
    name: ["John"],
  },
}, { runValidators: true, upsert: true, new: true }, function (err, data) {
  if (err) {
    return console.log(err);
  } else {
    // console.log(data.validateSync())
    return console.log('Updated', data);
  }
});
```

will allow you to update the user to have a `name` field of ["John"] without throwing any errors, even though the custom validator I included explicitly prohibits any name array of length less than or equal to 1. The validator itself works just fine, as can be seen by the fact that if you uncomment the line `console.log(data.validateSync())`, thereby forcing the validation to take place, it will in fact return the appropriate error message. The problem is that this validation is not taking place within the `findOneAndUpdate()` call, despite the fact that I include the `runValidators=true` option.

**Comment by vkarpov15** [comment]
Looks like a bug, can you open up a separate issue for that?

**Comment by rmacqueen** [comment]
Yep, opened https://github.com/Automattic/mongoose/issues/4039

**Comment by vkarpov15** [comment]
Thanks

**Comment by Saravanan90** [comment]
Is there any fix available for the above issue? @vkarpov15

**Comment by vkarpov15** [comment]
@Saravanan90 please stop commenting on long-closed issues without any meaningful information. Open up a separate issue with code samples.
